### PR TITLE
Adding source-hubspot-native connector

### PR DESCRIPTION
**Pull request by Luishfs**
**Description:**

Adding `source-hubspot-native` connector

**Documentation links affected:**
https://github.com/estuary/flow/pull/1440

**Notes for reviewers:**

Tests were made using a custom Hubspot account.
Each stream was tested one at a time with each stream. In order
to allow for the cursor to catch recent data, my `cut_off`/`start_date` variables
were set using ` - timedelta(days=365)` .

Pro streams data could not be validated. Because of that, they were passed on the base V3 
model which should handle all of their pagination cases. ( workflows needed a special pagination function
and resource).
Pro streams are:

- goal_targets
- feedback_submissions
- workflows

## Streams Interval

|Stream  | Interval  |
|--|--|
|Companies  | 30 seconds  |
|Contacts  | 30 seconds  |
|Deals  | 30 seconds  |
|Engagements  | 30 seconds  |
|Contact Lists  | 1 minute |
|Contact Lists Membership  | 1 minute  |
|Subscription Changes  | 1 minute  |
|Email Events  | 1 minute  |
|Ticket Pipelines  | 1 minute  |
|Deal Pipelines  | 1 minute  |
|Campaigns  | 1 minute  |
|Engagements Calls | 30 seconds  |
|Engagements Emails  | 30 seconds  |
|Engagements Tasks  | 30 seconds  |
|Engagements Notes | 30 seconds  |
|Goal Targets  | 30 seconds  |
|Line Items  | 30 seconds  |
|Tickets  | 30 seconds  |
|Email Subscriptions  | 30 seconds  |
|Marketing Forms  | 30 seconds  |
|Owners  | 30 seconds  |
|Properties  | 1 day  |
|Feedback submissions  | 30 seconds  |
|Marketing Emails  | 30 seconds  |
|Workflows  | 30 seconds  |
|All Custom objects  | 1 minute  |

Hubspot imposes a API rate limit of 100 requests per 10 seconds.
V3 Objects can require more than one request per stream, this happens because of the "Associations" and "batch" options, which uses more API requests than usual.
Because of that, a interval of 30 seconds was added, so that smaller streams ( like Companies or Campaigns ) don't add-up to the rate limit.

V1 Objects that can require a longer capture time were set to 60 seconds so that on general their capture can happen after the quicker V3 Object and dont trigger 429 responses that often. Since they can usually take hours to occur, real-time transfers are already blocked, but a longer wait time period can harm smaller clients that dont require long hours of capture.

Finally, Properties were set to 1 day. They are a really small stream that does not require real-time processing and wont affect usage.

- - -
This change is [Reviewable](https://reviewable.io/reviews/estuary/connectors/1438)

### CODE REVIEW

File: source-hubspot-native/test.flow.yaml
Comment (williamhbaker) [inline source-hubspot-native/test.flow.yaml:16]: What's the reason for adding this explicit interval to the test configuration?

File: source-hubspot-native/tests/snapshots/source_hubspot_native_tests_test_snapshots__capture__stdout.json
Comment (williamhbaker) [inline source-hubspot-native/tests/snapshots/source_hubspot_native_tests_test_snapshots__capture__stdout.json:5]: There's a lot changing in this snapshot. Understandably the new resources will add to it, but the churn on the pre-existing outputs is hard to digest. Is any of the existing output changing, or is just all from shifting the ordering of the things around?

File: source-hubspot-native/tests/snapshots/source_hubspot_native_tests_test_snapshots__discover__stdout.json
Comment (williamhbaker) [inline source-hubspot-native/tests/snapshots/source_hubspot_native_tests_test_snapshots__discover__stdout.json:3828]: `_meta/row_id` needs to be a required property for this to work well with most materializations. So `_meta` needs to be required,  and in `_meta` `row_id` needs to be required.

Technically there is a default value for it set so it might work without this, but I think we might as well set the field as required since we know it will always be there.

File: source-hubspot-native/tests/test_snapshots.py
Comment (williamhbaker) [inline source-hubspot-native/tests/test_snapshots.py:28]: I'm curious as to why we don't need to do this anymore?

File: source-hubspot-native/source_hubspot_native/resources.py
Comment (williamhbaker) [inline source-hubspot-native/source_hubspot_native/resources.py:8]: Why is this commented out?

File: source-hubspot-native/source_hubspot_native/resources.py
Comment (williamhbaker) [inline source-hubspot-native/source_hubspot_native/resources.py:165]: I asked about this elsewhere in a snapshot output as well - I'd like to know more about the default time delta of 7 minutes.

File: source-hubspot-native/source_hubspot_native/resources.py
Comment (williamhbaker) [inline source-hubspot-native/source_hubspot_native/resources.py:348]: The comment for `workflow_object` is nice - as I'm reading this I am wondering why `subscription_object` exists, so a similar comment explaining that would be appreciated.

File: source-hubspot-native/source_hubspot_native/resources.py
Comment (williamhbaker) [inline source-hubspot-native/source_hubspot_native/resources.py:425]: A comment (is that what they are called? docstrings is maybe the right term?) on this function would be nice too.

File: source-hubspot-native/source_hubspot_native/api.py
Comment (williamhbaker) [inline source-hubspot-native/source_hubspot_native/api.py:101]: A limit of 2 seems really small? Not sure if it was set at that before though.

File: source-hubspot-native/source_hubspot_native/api.py
Comment (williamhbaker) [inline source-hubspot-native/source_hubspot_native/api.py:362]: Can you comment more on what this function is for? Coming in without a deep understanding of the pre-existing code, I am struggling to understand why this is needed as an alternative to `fetch_changes`. I see that it does things a little differently and that it doesn't have that final `fetch_batch_with_associations` part, but I need a little help understanding what the reasoning for that is.

File: source-hubspot-native/source_hubspot_native/api.py
Comment (williamhbaker) [inline source-hubspot-native/source_hubspot_native/api.py:491]: Other similar seeming functions use `result.has_more and result.time_offset` as the return value for the `PageCursor`, seemingly as an alternative for this if/else. Would that work for these too?

File: source-hubspot-native/source_hubspot_native/api.py
Comment (williamhbaker) [inline source-hubspot-native/source_hubspot_native/api.py:479]: We're using `{"count": 1}` elsewhere if there is no pre-existing page cursor. Why 250 here?

File: source-hubspot-native/source_hubspot_native/api.py
Comment (williamhbaker) [inline source-hubspot-native/source_hubspot_native/api.py:275]: Is there not a way to pass the timestamp parameter anymore? Do we need to filter the result on `since` as we will now I assume get every single ticket each time?

File: source-hubspot-native/source_hubspot_native/api.py
Comment (williamhbaker) [inline source-hubspot-native/source_hubspot_native/api.py:494]: This filtering conceptually makes since to me...since we want things after `since`, only emit documents that come after `since`. But I don't know why we don't have that on others, like `fetch_recent_contacts`? It seems they are doing pretty much the same thing. How does that work for `fetch_recent_contacts`, but it wouldn't work here?

**Review by williamhbaker — COMMENTED** [review]
Made some initial comments!

File: source-hubspot-native/test.flow.yaml
Comment (Luishfs) [reply source-hubspot-native/test.flow.yaml:16]: If i got it right, when a user does not generate their own bindings, we generate then by the default `common.Resource` resource, and the default wait value for estuary is 7 minutes. That's why this specific value at this place ( since discovery would see that no interval was set, i've added that ) .
Does that make sense?

File: source-hubspot-native/tests/snapshots/source_hubspot_native_tests_test_snapshots__capture__stdout.json
Comment (Luishfs) [reply source-hubspot-native/tests/snapshots/source_hubspot_native_tests_test_snapshots__capture__stdout.json:5]: Lots of shifiting around since this snapshot was first added when we had only 4 streams

File: source-hubspot-native/tests/snapshots/source_hubspot_native_tests_test_snapshots__discover__stdout.json
Comment (Luishfs) [reply source-hubspot-native/tests/snapshots/source_hubspot_native_tests_test_snapshots__discover__stdout.json:3828]: To solve the second requirement, i did a modification at estuary-cdk in this commit 344be6d90aa7b0f286f210242c7afa9f77c1279d

The first one ( add `_meta/row_id` ) seem a bit harder since its already required by the `BaseDocument` model, and i don't really know how to make both of then required since the models listed in models.py already require the BaseDocument model, could you help me to solve it?

File: source-hubspot-native/source_hubspot_native/resources.py
Comment (Luishfs) [reply source-hubspot-native/source_hubspot_native/resources.py]: This was for code standardization, i saw that google-sheets-native was using `from estuary_cdk.capture import common` and decided that it would be best to keep everything stardard from this initial point forward, nothing important

File: source-hubspot-native/source_hubspot_native/resources.py
Comment (Luishfs) [reply source-hubspot-native/source_hubspot_native/resources.py:164]: I believe to have answered that on one of the more initial comments, please tell me if that explanation is enough

File: source-hubspot-native/source_hubspot_native/resources.py
Comment (Luishfs) [reply source-hubspot-native/source_hubspot_native/resources.py:347]: A comment for this specific function was added here a484280098b57da301ffc5bc09ab89fa835fc8d0

File: source-hubspot-native/source_hubspot_native/resources.py
Comment (Luishfs) [reply source-hubspot-native/source_hubspot_native/resources.py:437]: A comment for this specific function was added here a484280098b57da301ffc5bc09ab89fa835fc8d0

File: source-hubspot-native/source_hubspot_native/api.py
Comment (Luishfs) [reply source-hubspot-native/source_hubspot_native/api.py:503]: They are being handled by the fetch_changes_no_batch function, more on how it works in their docstring added here a484280098b57da301ffc5bc09ab89fa835fc8d0

File: source-hubspot-native/source_hubspot_native/api.py
Comment (Luishfs) [reply source-hubspot-native/source_hubspot_native/api.py:491]: Hubspot limits the amount of lists by 250 ( if you have Pro/Enterprise it seems to handle more), and data from this endpoint is really small. If no page is given, instead of creating a pagination object with 100 data and have it pass 3 times, i just saw more simpler to handle all 250 lists at once.

If there are more lists, pagination will occur normally

File: source-hubspot-native/source_hubspot_native/api.py
Comment (Luishfs) [reply source-hubspot-native/source_hubspot_native/api.py:275]: So, when developing this i had 2 main thoughts in my head ( Hubspot [docs](https://legacydocs.hubspot.com/docs/methods/tickets/get-ticket-changes) for reference):

1. If im using this endpoint to trigger which tickets have changed, i will not get all tickets. They do not specify what is a change, and i dont think a  ticket creation would reach this.
2. Even if it did, when i select the timestamp i would never allow for a """full_refresh""" pattern where i get all data. From what i understood, the since is a way of us understading where the last data parsed was set in history, not as means of getting a specific timeframe

Comment (Luishfs) [reply source-hubspot-native/source_hubspot_native/api.py:275]: > 2\. the `since` is a way of us understading where the last data parsed was set in history, not as means of getting a specific timeframe

just further explaining this quote, i believe the `timestamp` parameter would return to me 'timestamp - 24 >= data => timestamp`  as is cited in the docs

File: source-hubspot-native/source_hubspot_native/api.py
Comment (Luishfs) [reply source-hubspot-native/source_hubspot_native/api.py:491]: This is a really clever idea from johnny, were we call this [endpoint](https://legacydocs.hubspot.com/docs/methods/lists/get_list_contacts_recent)  where it formats the results in order of more recent to less recent. Then, the `fetch_recent_contacts` method gets when this data has been updated and its Id and sents it to the `fetch_changes` function. This function is responsible to filter these timestamps with our `cut_off` / `since` value, and the ones that passes are then collected using batch functions

File: source-hubspot-native/tests/test_snapshots.py
Comment (Luishfs) [reply source-hubspot-native/tests/test_snapshots.py:28]: Im not sure what this did, but it was made to remove these specific timestamps specifically from v3 objects properties, i didnt deem it necessary as there were no comments explaining why ( my idea was that this was probably breaking tests? )

File: source-hubspot-native/source_hubspot_native/api.py
Comment (Luishfs) [reply source-hubspot-native/source_hubspot_native/api.py]: Yes, raised in 0b549e9f4cdfafdc3aadf8255959108dfb071aab to be the default value. Will leave johnny comments for future reference in case the limit number needs to be revised

**Comment by Luishfs** [comment]
@williamhbaker  I believe i've answer all the initial comments!

File: source-hubspot-native/tests/snapshots/source_hubspot_native_tests_test_snapshots__discover__stdout.json
Comment (williamhbaker) [reply source-hubspot-native/tests/snapshots/source_hubspot_native_tests_test_snapshots__discover__stdout.json:3828]: I don't think we want `row_id` to be required in `_meta` across the board for all connectors, since some of them won't have a concept of a `row_id` and will just use `_meta` to report `op`.

So, given that there is not a straightforward way to set this field as required and it _does_ have a default value which will allow it to be materialized (I actually tested this with our `source-google-sheets-native` connector, which has the same kind of schema), I'll retract my statement that this needs to be changed, and we should leave it alone for now and back out the change in https://github.com/estuary/connectors/commit/344be6d90aa7b0f286f210242c7afa9f77c1279d

File: source-hubspot-native/tests/test_snapshots.py
Comment (williamhbaker) [reply source-hubspot-native/tests/test_snapshots.py:28]: OK yeah, I guess if the tests work repeatedly without it then it shouldn't be a problem.

File: source-hubspot-native/source_hubspot_native/resources.py
Comment (williamhbaker) [reply source-hubspot-native/source_hubspot_native/resources.py]: OK, having it removed entirely seems good, just didn't want to leave a mysterious line of commented code in. But it looks good now.

File: source-hubspot-native/test.flow.yaml
Comment (williamhbaker) [reply source-hubspot-native/test.flow.yaml:16]: Yeah, that does make sense. I am wondering how/why these bindings were originally set to have no explicit `interval`. I don't 100% know what happens in that case, and have raised the question in the #saas-connectors slack channel.

And 7 minutes is a pretty weird value, lol. We use 30 seconds in a lot of other places, or maybe 5 minutes, but I'm not sure about a default wait value for estuary of 7 minutes?

Comment (Luishfs) [reply source-hubspot-native/test.flow.yaml:16]: I dont know why but in my head 7 minutes was the default interval. Really don't remember where that came from ( probably need my meds ). @williamhbaker Should i leave the default interval of 30 seconds?

File: source-hubspot-native/tests/snapshots/source_hubspot_native_tests_test_snapshots__discover__stdout.json
Comment (Luishfs) [reply source-hubspot-native/tests/snapshots/source_hubspot_native_tests_test_snapshots__discover__stdout.json:3828]: Got it!
Changes were reverted here 051e02e8056c6f40d56685a485d3ba9fc8674bed

File: source-hubspot-native/test.flow.yaml
Comment (williamhbaker) [reply source-hubspot-native/test.flow.yaml:16]: 30 seconds seems like a reasonable default interval for bindings that do not otherwise have restrictive rate limits, and are incremental in nature. We wouldn't want to re-fetch a huge snapshot every 30 seconds for example, so bindings like that should have larger intervals. I'll leave it to you to decide on the intervals based on your knowledge of how hubspot works, since I'm not very familiar with it other than these general statements.

Comment (williamhbaker) [reply source-hubspot-native/test.flow.yaml:16]: Another note is that I don't know why these pre-existing bindings had no interval at all, and were set to just re-fetch as fast as possible. Maybe there's a good reason for that?

Comment (dyaffe) [reply source-hubspot-native/test.flow.yaml:16]: I think because hubspot lets you and we can get real-time updates this way.

Comment (Luishfs) [reply source-hubspot-native/test.flow.yaml:16]: @dyaffe @williamhbaker  Hubspot limits us by 100 requests every 10 seconds, and given that we have 27+ streams, i think having a 30+ seconds breather for each stream would really help to not reach the limit. I will test with some values and get more data on this matter

Comment (Luishfs) [reply source-hubspot-native/test.flow.yaml:16]: @williamhbaker Added intervals in 4171ca4368a0cd29e8af7b77533686cb6c80a249

Comment (williamhbaker) [reply source-hubspot-native/test.flow.yaml:16]: I'm still seeing this `PT420S` interval in `test.flow.yaml`, which I don't think is what we want.

Comment (Luishfs) [reply source-hubspot-native/test.flow.yaml]: @williamhbaker Forgot to run discover, rebased the commit here da3558d73fb91d5fa30662cc2462b84f6fe451fb

Comment (Luishfs) [reply source-hubspot-native/test.flow.yaml]: @williamhbaker Forgot to run discover, rebased the commit here fda33783a560759e3bdc40251a9633277d602c9f

**Comment by Luishfs** [comment]
@williamhbaker Tests are now pasing

File: source-hubspot-native/test.flow.yaml
Comment (dyaffe) [reply source-hubspot-native/test.flow.yaml]: @Luishfs just one more push here.  We spoke about having 27 streams and some of them have 3 api calls per stream.  The worst case scenario seems to me that every single one has 3 calls.  If that's the case, we could lower the limit from 30s to 10s with room to spare since 10s would be at max 81 API requests every 10s.

Let me know if I'm wrong for any reason.

File: source-hubspot-native/test.flow.yaml
Comment (williamhbaker) [inline source-hubspot-native/test.flow.yaml:36]: Just spot checking: The table in the PR description says that this is supposed to be 1 day. But its 30 seconds here. Which one is it supposed to be?

**Review by williamhbaker — APPROVED** [review]
LGTM. See comment about interval on one of the streams. I didn't go through and check every single one line by line against the table in the PR description, but I did find that discrepancy at a glance. I trust that you'll make sure things match up before merging.

File: source-hubspot-native/test.flow.yaml
Comment (Luishfs) [reply source-hubspot-native/test.flow.yaml:36]: Updated at ae65e51c85616092acd08ece0c3ace515985e4d3

**Review by dyaffe — APPROVED** [review]

**Pull request merged**
